### ISSUE TITLE
Are backend changes thread-safe? Should they be?

### ISSUE THREAD

**Issue by hamogu** [sherpa/sherpa#1879]
I recently added the ability to change I/O and plotting backends. In both cases, I wired that into the existing Sherpa infrastructure by essentially monkey-patching `sherpa.io.backend` and `sherpa.plot.backend` when a backend is changed. That's fine for the intended purpose, i.e. some light local testing  or non-threaded test execution (as we do in CI) and in particular, it's fine for interactive use.

One might assume that most batched, scripted use of Sherpa selects one backend and sticks with it, so this might not be a big issue in practice.

However, over the long run, I'm concerned that we bake in a non thread-safe design in a world that increasingly moves to multiprocessing and multi-threading. For now, a warning in the docs is probably enough, but do we need a redesign in the future?

One way to do that would be to explicitly pass the a backend to an I/O or plotting function (with `None` meaning "currently set backend" as default?).

**Comment by DougBurke** [comment]
It's not obvious to me that just explicitly sending in a backend is enough to solve all possible tread safety issues (e.g. is it possible for the same backend object is being used in different contexts/threads/whatevers). So I'm wary about designing an API that is meant to be "thread safe" without a better understanding of the issues (which currently do not have). As we move towards a GIL-less world I think there will be more documentation/discussion/... on things we need to worry about, so although it would be nice to "get it right" first time, I am open to us having to make changes in later releases (for some definition of "amount of needed work" and "change" and "open" which I am not going to define so I can ignore this comment in the future...).

**Comment by hamogu** [comment]
I agree with that assessment, but at least we now have an issue to point to if someone asks "why did you not..."
